### DISCUSSION TITLE
Creating a duplicate database object for the same connection - using jest.mock

### DISCUSSION THREAD

**Discussion by tomskopek** [vitaly-t/pg-promise#851] [Q&A]
When I use the database object in a test, *and in the same test, I mock a helper function which also uses the database*, I get the error:

`Creating a duplicate database object for the same connection`

Any idea why this is?

Here is a minimal reproducible example:
https://github.com/tomskopek/pg-promise-duplicate-database-object-jest-mock-example

**Comment by vitaly-t** [comment]
See [here](https://stackoverflow.com/questions/34382796/where-should-i-initialize-pg-promise#34427278).

If your issue is however specifically with tests, you can deactivate all warnings, using [`noWarnings` flag](http://vitaly-t.github.io/pg-promise/module-pg-promise.html).

**Reply by tomskopek** [reply]
Sorry, I don't think I was clear enough in my question!

I'm aware of that pattern of putting it in it's own module (and that's what we're doing). And to be clear, I'm fairly certain the singleton pattern would fix the issue here. We haven't implemented the singleton pattern because it doesn't seem to be an issue for us.

This question is more of a curiosity about jest and mock:

I'm wondering why this error gets thrown with this mock syntax:
```js
jest.mock('../sum')
```

but not this one:
```js
jest.mock('../sum', () => {
    return () => 3
})
```

If you look at the reproducible example, instructions should be fairly clear, but to save the trouble, here are relevant files:

```js
// sum.js
const { db } = require('./database')

async function sum(a, b) {
    await db.oneOrNone('SELECT * FROM information_schema.tables LIMIT 1'); // <------ arbitrary db operation
    return a + b;
}

module.exports = sum;
```

```js
// sum.test.js
const sum = require('../sum')
const { db } = require('../database')

// 1. run the test, should have no error

// 2. Uncomment this mock, run `yarn test`, see how it throws "Creating a duplicate database object for the same connection"
// jest.mock('../sum')

// 3. Comment the above mock, uncomment this mock, see how it doesnt throw that same error
// jest.mock('../sum', () => {
//     return () => 3
// })

describe('sum', () => {
    beforeEach(()=>{
        // some setup for the integration test
        db.oneOrNone('SELECT * FROM information_schema.tables LIMIT 1'); // <----- arbitrary db operation
    })
    it('should return 2 when given 1 and 1', async () => {
        const result = await sum(1,1)
        expect(result).toEqual(2)
    });
});
```

```js
// database.js
const initOptions = {};
const pgp = require('pg-promise')(initOptions);

const connection = {
    "user": "dev",
    "password": "password",
    "host": "localhost",
    "port": 5432,
    "database": "test_duplicate_connections"
}

const db = pgp(connection)

module.exports = { db, pgp }
```

**Comment by vitaly-t** [comment]
I'm not entirely sure how that `mock` works, but I suspect that in the first case it executes with what's inside the function, while in the second one it replaces the function body, so no queries ever execute. Just a guess.

**Reply by tomskopek** [reply]
Right, but even so, running the test without any mocks at all executes with what's inside the function, right? But that doesn't produce an error. So wondering what's different about the mock.

Anyways, thanks!
